bundles/org.eclipse.ui.ide/extensions/org/eclipse/ui/actions/WorkspaceModifyOperation.java

```diff
@@ -114,6 +114,13 @@ public synchronized final void run(IProgressMonitor monitor)
 				} catch (InterruptedException e2) {
 					// Re-throw as OperationCanceledException, which will be
 					// caught and re-thrown as InterruptedException below.
+
+					// This is the best case scenario because the original exception has the exact
+					// line where the cancelation happened (better for debugging)
+					if (e2.getCause() instanceof OperationCanceledException oce)
+						throw oce;
+
+					// fallback: create a new exception
```

Comment:
Why has this exception to be repackaged outside the workspace runnable at all? The current handling is broken, as the catch block for the execution of the workspace runnable adds _every_ `OperationCanceledException` as cause to an `InterruptedException` although the `OperationCanceledException` is only used as temporary non-checked exception wrapped (see comment in lines 115--116), so this is not suppoed to be passed to the method consumer.

From my understand, what this code is simply trying to do is to pass the `InterruptedException` outside of the runnable and throw it there. So the handling could be simply done like for the invocation target exception in line 113.

Reply:
> From my understand, what this code is simply trying to do is to pass the InterruptedException outside of the runnable and throw it there.

That was my understanding too, the thing is that simply storing the exception in `iteHolder` like it's done in line 113 would not skip the execution of the rest of the code outside the inner `try-catch`-block while throwing an `OperationCanceledException` would.
The objective of this PR is simply adding (in this particular case: **improving**) details to the `InterruptedException`, not to change the logic.

Another option to preserve the logic would be to check if `iteHoldler` holds an `OperationCanceledException` and, if it does, skip processing all pending _asyncExecs_, but that code wouldn't look much better.

Reply:
> That was my understanding too, the thing is that simply storing the exception in `iteHolder` like it's done in line 113 would not skip the execution of the rest of the code outside the inner `try-catch`-block while throwing an `OperationCanceledException` would.

Can you please detail that? I do not see any further code that would be executed then.

Reply:
If line 113 is reached then this piece of code will still be invoked:
```java
// if we are in the UI thread, make sure we use progress monitor
// that spins event loop to allow processing of pending asyncExecs
if (monitor != null && PlatformUI.isWorkbenchRunning()
		&& !PlatformUI.getWorkbench().isStarting()) {
	Display display = PlatformUI.getWorkbench().getDisplay();
	if (!display.isDisposed()
			&& display.getThread() == Thread.currentThread()) {
		monitor = new EventLoopProgressMonitor(monitor);
	}
}
IDEWorkbenchPlugin.getPluginWorkspace().run(workspaceRunnable,
		rule, IResource.NONE, monitor);
```

But if an exception is thrown then that piece of code will not run.

Reply:
That code is always invoked, as line 113 is within a lambda that is triggered in line 138.

Reply:
You're right, thanks. That means I can safely use the same mechanism in the whole method. 

Done in b6c33cdd30a6d9eb63686e48173e7322cb17f6cf